### PR TITLE
Raw packet mode and interceptor

### PR DESCRIPTION
**Pull request by stepheny**
Encoder and decoder were slightly modified to accept rawPacket option, which bypasses ogg page handling. Opus packet relies on underlying framing mechanics, ogg is good choice but we have websocket.
Encoder and decoder should behave the same without rawPacket option.
Interceptor example can be easily modified to hook websocket with either rawPacket or ogg page.

I've added '_exp2' to DEFAULT_EXPORTS in Makefile. I've built emscripten-fastcomp from source. Without special config, emcc with --llvm-lto 1 removes _exp2.

Compiled scripts in dist were not updated. Maybe you would want to update them yourself.
And you could checkout my raw_packet_dist branch for a quick look.

### CODE REVIEW

**Comment by chris-rudmin** [comment]
Thanks for the new pull request! It's not clear what `interceptor.js` is for.  Also I don't see why `_exp2` needs to be exported. I don't see it used anywhere. 

I would support a raw opus option, but I think the implementation should be separated from the `encoderWorker.js` code. For example: we could have `opusEncoderWorker.js`, `opusDecoderWorker.js`, `oggMuxer.js`, `oggDemuxer.js`, so that the functionality can be separated and to avoid code duplication. What do you think?

**Comment by chris-rudmin** [comment]
Ah, i just read more carefully your comment about _exp2.  I'll see what happens when run `make` :)

**Comment by stepheny** [comment]
The interceptor is for streaming. I knew it is quite different from Recorderjs' original idea. Maybe you could just reject it or throw it into a separate branch.
In brief interceptor works as both a capture and a renderer, tightly coupled with encoder and decoder. I'm trying to implement some VoIP stuff. Latency would be critical if it is a bidirectional conversation.Maybe I should try to use opus directly, but I don't think I could do it better than what Recorderjs did.
Raw opus packet is also for latency. Due to frequent packets, it did suffer a great traffic penalty to the original ogg page mode, less payload but more traffic. But it achieves great codec delay. If we config ogg to use a 1ms page, a similar delay could be achieved, but in that case ogg has more overhead.

I think it would be great if the codec and muxer were designed to be separated at the very beginning. But in current state, I doubt it is worthy to  refactor almost whole project. I cannot imagine a future of porting a codec framework into javascript. It should be done by system library or at least the browser.

The dist library delivered in you repo do not lack _exp2 symbol. It could be something went wrong in my emscripten toolchain, but I cannot figure out where. At lease, it should not harm to explicit export _exp2.
In my case, the dist lacking _exp2, would work the same if encoder bitrate is set to 64000 or more. If bitrate is set below 64000, there would be a great chance to run into an abort.

**Comment by chris-rudmin** [comment]
I'd be happy to merge this with a few changes, I'll add comments into the pull request.

File: Makefile [hunk 1]
```diff
@@ -21,11 +21,14 @@ LIBSPEEXDSP_EXPORTS:='_speex_resampler_init','_speex_resampler_process_interleav
...
 
+INTERCEPTOR=$(OUTPUT_DIR)/interceptor.min.js
+INTERCEPTOR_SRC=$(INPUT_DIR)/interceptor.js
+
 WAVE_WORKER=$(OUTPUT_DIR)/waveWorker.min.js
 WAVE_WORKER_SRC=$(INPUT_DIR)/waveWorker.js
 
 
-default: $(LIBOPUS_ENCODER) $(LIBOPUS_DECODER) $(RECORDER) $(WAVE_WORKER) test
+default: $(LIBOPUS_ENCODER) $(LIBOPUS_DECODER) $(RECORDER) $(INTERCEPTOR) $(WAVE_WORKER) test
```
Comment (chris-rudmin) [inline Makefile]: Adding the interceptor.js does not seem necessary, as the changes have been made to `recorder.js`

File: example/interceptor.html [hunk 2]
```diff
@@ -0,0 +1,226 @@
+<!DOCTYPE html>
```
Comment (chris-rudmin) [inline example/interceptor.html]: I would prefer one example.html.  Any new options could be moved into the existing example.

File: Makefile (hunk 1 above)
Comment (chris-rudmin) [inline Makefile]: `Interceptor.js` is not really necessary, as any relevant changes have been made to `recorder.js`

File: src/recorder.js [hunk 3]
```diff
@@ -26,6 +26,7 @@ var root = (typeof self === 'object' && self.self === self && self) || (typeof g
     this.config.encoderSampleRate = config.encoderSampleRate || 48000;
     this.config.encoderPath = config.encoderPath || 'encoderWorker.min.js';
     this.config.streamPages = config.streamPages || false;
+    this.config.rawPacket = config.rawPacket || false;
```
Comment (chris-rudmin) [inline src/recorder.js]: Please add this new config option and some documentation to the readme.

File: src/recorder.js [hunk 4]
```diff
@@ -148,7 +149,13 @@ var root = (typeof self === 'object' && self.self === self && self) || (typeof g
       var that = this;
       this.encoder = new global.Worker( this.config.encoderPath );
 
-      if (this.config.streamPages){
+      if (this.config.rawPacket){
```
Comment (chris-rudmin) [inline src/recorder.js]: Less code duplication if we did `if (this.config.rawPacket || this.config.streamPages) { ...` Also please add a note this behaviour in the readme.

File: src/encoderWorker.js [hunk 5]
```diff
@@ -101,7 +102,7 @@ var root = (typeof self === 'object' && self.self === self && self) || (typeof g
   OggOpusEncoder.prototype.encodeFinalFrame = function() {
     var finalFrameBuffers = [];
     for ( var i = 0; i < this.numberOfChannels; ++i ) {
-      finalFrameBuffers.push( new Float32Array( this.bufferLength - (this.resampleBufferIndex / this.numberOfChannels) ));
+      finalFrameBuffers.push( new Float32Array( (this.resampleBufferLength - this.resampleBufferIndex) / this.numberOfChannels ));
```
Comment (chris-rudmin) [inline src/encoderWorker.js]: Could you add a unit test for this bug fix?

File: src/decoderWorker.js [hunk 6]
```diff
@@ -30,10 +30,31 @@ var OggOpusDecoder = function( config ){
   this.decoderSampleRate = config['decoderSampleRate'] || 48000;
   this.outputBufferSampleRate = config['outputBufferSampleRate'] || 48000;
   this.resampleQuality = config['resampleQuality'] || 3;
+  this.rawPacket = config['rawPacket'] || false;
   this.outputBuffers = [];
+  if (this.rawPacket){ this.inited = false };
 };
 
 OggOpusDecoder.prototype.decode = function( typedArray ) {
+  if (this.rawPacket) {
```
Comment (chris-rudmin) [inline src/decoderWorker.js]: I'm a fan of one return point for each method. IMO a code block of this size should be broken into two. Perhaps `OggOpusDecoder.prototype.decodeRawPacket` and `OggOpusDecoder.prototype.decodeOggContainer` or something like that.

File: src/recorder.js (hunk 3 above)
Comment (chris-rudmin) [inline src/recorder.js]: Also please add this new config option to the unit tests.

File: src/encoderWorker.js [hunk 7]
```diff
@@ -252,6 +256,14 @@ var root = (typeof self === 'object' && self.self === self && self) || (typeof g
   OggOpusEncoder.prototype.segmentPacket = function( packetLength ) {
     var packetIndex = 0;
 
+    if (this.rawPacket)
```
Comment (chris-rudmin) [inline src/encoderWorker.js]: Would love to see unit tests for new `rawPacket` behaviour.

File: src/decoderWorker.js [hunk 8]
```diff
@@ -30,10 +30,31 @@ var OggOpusDecoder = function( config ){
...
+    if (!this.inited) {
```
Comment (chris-rudmin) [inline src/decoderWorker.js]: Naming, could be confusing because this is only used for rawPackets. Perhaps change the name to `rawPacketInitialized`

**Pull request closed without merging**